glommio/src/sys/blocking.rs

```diff
@@ -110,35 +112,99 @@ pub(super) struct BlockingThreadResp {
 
 pub(super) type BlockingThreadHandler = Box<dyn Fn(BlockingThreadResult)>;
 
-pub(super) struct BlockingThread {
-    eventfd: RawFd,
-    queue: Arc<ArrayQueue<BlockingThreadReq>>,
-    responses: Arc<ArrayQueue<BlockingThreadResp>>,
-    waiters: RefCell<BTreeMap<u64, BlockingThreadHandler>>,
+struct BlockingThread(JoinHandle<()>);
+
+impl BlockingThread {
+    pub(super) fn new(
+        reactor_sleep_notifier: Arc<SleepNotifier>,
+        rx: Arc<Receiver<BlockingThreadReq>>,
+        tx: Arc<Sender<BlockingThreadResp>>,
+        bindings: Option<impl IntoIterator<Item = usize> + Send + 'static>,
+    ) -> Self {
+        Self(std::thread::spawn(move || {
+            if let Some(bindings) = bindings {
+                bind_to_cpu_set(bindings).expect("failed to bind blocking thread");
+            }
+            while let Ok(el) = rx.recv() {
+                let res = el.op.execute();
+                let id = el.id;
+                let resp = BlockingThreadResp { id, res };
+
+                if tx.send(resp).is_err() {
+                    panic!("failed to send response");
+                }
+                reactor_sleep_notifier.notify(false);
+            }
+        }))
+    }
+}
+
+pub(crate) struct BlockingThreadPool {
+    tx: Sender<BlockingThreadReq>,
+    rx: Receiver<BlockingThreadResp>,
+    waiters: RefCell<AHashMap<u64, BlockingThreadHandler>>,
     requests: Cell<u64>,
+    _threads: Vec<BlockingThread>,
 }
 
-impl BlockingThread {
-    pub(super) fn push(&self, op: BlockingThreadOp, action: Box<dyn Fn(BlockingThreadResult)>) {
+impl BlockingThreadPool {
+    pub(crate) fn new(
+        placement: PoolPlacement,
+        sleep_notifier: Arc<SleepNotifier>,
+    ) -> io::Result<Self> {
+        let (in_tx, in_rx) = crossbeam::channel::unbounded();
```

Comment:
can we use bounded channels?

Reply:
Ideally I would like to find a channel that would have an async TX and a sync RX. 

I don't think it is okay to panic if the que is full like we did before. 

And it's also not okay to bubble up blocking to user code. This is meant to be async.

Reply:
So an ideal way to deal with this would be to provide async back pressure with an async TX. I haven't found one though. 

I think there is a way to use tokio channels that way. I'll look into it.

Reply:
See if this does what you need: 
https://docs.rs/async-channel/latest/async_channel/

Reply:
It doesn't. Tokio channel however does have a `blocking_recv()`, but that would imply us depending on Tokio. Going to merge as-is but I agree this is something we'll need to keep an eye on.